text/0000-async-stream.md

```diff
@@ -0,0 +1,792 @@
+
+// First, the struct:
+
+/// A stream which counts from one to five
+struct Counter {
+    count: usize,
+}
+
+// we want our count to start at one, so let's add a new() method to help.
+// This isn't strictly necessary, but is convenient. Note that we start
+// `count` at zero, we'll see why in `poll_next()`'s implementation below.
+impl Counter {
+    fn new() -> Counter {
+        Counter { count: 0 }
+    }
+}
+
+// Then, we implement `Stream` for our `Counter`:
+
+impl Stream for Counter {
+    // we will be counting with usize
+    type Item = usize;
+
+    // poll_next() is the only required method
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Increment our count. This is why we started at zero.
+        self.count += 1;
+
+        // Check to see if we've finished counting or not.
+        if self.count < 6 {
+            Poll::Ready(Some(self.count))
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+```
+
+## Initial impls
+
+There are a number of simple "bridge" impls that are also provided:
+
+```rust
+impl<S> Stream for Box<S>
+where
+    S: Stream + Unpin + ?Sized,
+{
+    type Item = <S as Stream>::Item
+}
+
+impl<S> Stream for &mut S
+where
+    S: Stream + Unpin + ?Sized,
+{
+    type Item = <S as Stream>::Item;
+}
+
+impl<S, T> Stream for Pin<P>
+where
+    P: DerefMut<Target=T> + Unpin,
+    T: Stream,
+{
+    type Item = <T as Stream>::Item;
+}
+
+impl<S> Stream for AssertUnwindSafe<S>
+where
+    S: Stream, 
+{
+    type Item = <S as Stream>::Item;
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section goes into details about various aspects of the design and
+why they ended up the way they did.
+
+## Where does `Stream` live in the std lib?
+
+`Stream` will live in the `core::stream` module and be re-exported as `std::stream`.
+
+It is possible that it could live in another area as well, though this follows
+the pattern of `core::future`.
+
+## Why use a `poll` method?
+
+An alternative design for the stream trait would be to have a trait
+that defines an async `next` method:
+
+```rust
+trait Stream {
+    type Item;
+    
+    async fn next(&mut self) -> Option<Self::Item>;
+}
+```
+
+Unfortunately, async methods in traits are not currently supported,
+and there [are a number of challenges to be
+resolved](https://rust-lang.github.io/wg-async-foundations/design_notes/async_fn_in_traits.html)
+before they can be added. 
+
+Moreover, it is not clear yet how to make traits that contain async
+functions be `dyn` safe, and it is important to be able to pass around `dyn
+Stream` values without the need to monomorphize the functions that work
+with them.
+
+Unfortunately, the use of poll does mean that it is harder to write
+stream implementations. The long-term fix for this, discussed in the [Future possiblilities](future-possibilities) section, is dedicated [generator syntax].
```

Comment:
Broken link

Reply:
any chance you can open a PR with fixes?